File: pyorchestrator/backend/app/utils/semver.py
import re


def parse_version(tag: str) -> str:
    return re.sub(r"^v", "", tag.strip(), flags=re.IGNORECASE)


def _parts(version: str) -> list[int]:
    parts: list[int] = []
    for chunk in re.split(r"[.\-+]", version):
        if chunk.isdigit():
            parts.append(int(chunk))
        elif chunk:
            parts.append(0)
    return parts or [0]


def is_newer_version(candidate: str, current: str) -> bool:
    a = _parts(parse_version(candidate))
    b = _parts(parse_version(current))
    length = max(len(a), len(b))
    a.extend([0] * (length - len(a)))
    b.extend([0] * (length - len(b)))
    return a > b

File: dashboard/src/utils/semver.ts
export function parseVersion(tag: string): number[] {
  return tag
    .replace(/^v/i, '')
    .split(/[.+_-]/)
    .map((part) => {
      const n = parseInt(part.replace(/[^0-9].*$/, ''), 10);
      return Number.isFinite(n) ? n : 0;
    });
}

export function isNewerVersion(latest: string, current: string): boolean {
  const a = parseVersion(latest);
  const b = parseVersion(current);
  const len = Math.max(a.length, b.length);
  for (let i = 0; i < len; i += 1) {
    const av = a[i] ?? 0;
    const bv = b[i] ?? 0;
    if (av > bv) return true;
    if (av < bv) return false;
  }
  return false;
}
